src/client/networkclient.cc

```diff
@@ -0,0 +1,410 @@
+          << "    code: " << mpe->err->code.c_str()
+          << "  msg: " << mpe->err->msg.c_str() << endl;
+    }
+
+    disconnect();
+  }
+}
+
+void NetworkClient::handleConnectionsMessage(msg_ptr msg) {
+  // TODO - is this something that client should implement in a subclass?
+}
+
+void NetworkClient::handleRegistryReplyMessage(msg_ptr msg) {
+  // TODO - is this something that client should implement in a subclass?
+}
+
+void NetworkClient::handleMthdResMessage(msg_ptr msg) {
+  // TODO - is this something that client should implement in a subclass?
+}
+
+void NetworkClient::handlePluginTriggerRunMessage(msg_ptr msg) {
+  // TODO - is this something that client should implement in a subclass?
+}
+
+void NetworkClient::handleConnErrorMessage(msg_ptr msg) {
+  messages::MsgConnectionError* cem
+      = dynamic_cast<messages::MsgConnectionError*>(msg.get());
+  if (cem) {
+    if (output()) {
+      *output()
+          << "Received connection error message. Aborting connection..."
+          << endl << "    code: " << cem->err->code.c_str()
+          << "  msg: " << cem->err->msg.c_str() << endl;
+    }
+
+    disconnect();
+  }
+}
+
+void NetworkClient::handlePluginMethodRunMessage(msg_ptr msg) {
+  // TODO - is this something that client should implement in a subclass?
+}
+
+void NetworkClient::handlePluginQueryGetMessage(msg_ptr msg) {
+  // TODO - is this something that client should implement in a subclass?
+}
+
+void NetworkClient::handleBroadcastRunMessage(msg_ptr msg) {
+  // TODO - is this something that client should implement in a subclass?
+}
+
+void NetworkClient::handlePluginHandlerUnregisteredMessage(msg_ptr msg) {
+  // TODO - is this something that client should implement in a subclass?
+}
+
+void NetworkClient::setConnectionStatus(ConnectionStatus connection_status) {
+  if (status_ != connection_status) {
+    status_ = connection_status;
+    emit connectionStatusChanged(status_);
+    if (output()) {
+      *output() << "NetworkClient: New connection status: "
+          << connStatusStr(connection_status) << "." << endl;
+    }
+  }
+}
+
+void NetworkClient::socketConnected() {
+  if (output()) {
+    *output() << "NetworkClient: TCP socket connected - sending an "
+        "authentication key and \"connect\" message." << endl;
+  }
+
+  node_tree_ = new NodeTree(this);
+  client_socket_->write(authentication_key_);
+  sendMsgConnect();
+}
+
+void NetworkClient::socketDisconnected() {
+  setConnectionStatus(ConnectionStatus::NotConnected);
+  if(output()) {
+    *output() << "NetworkClient: TCP socket disconnected." << endl;
+  }
+
+  if (node_tree_) {
+    delete node_tree_;
+    node_tree_ = nullptr;
+  }
+
+  if (client_socket_) {
+    client_socket_->deleteLater();
+    client_socket_ = nullptr;
+  }
+}
+
+void NetworkClient::newDataAvailable() {
+  while (client_socket_) {
+    if (msg_ptr msg =
+        msgpack_wrapper_.loadMessage(client_socket_)) {
```

Comment:
Add some try-catch for `SchemaError` to avoid crash if we get bad message from server?